packages/codecs-core/src/codec.ts

```diff
@@ -4,36 +4,63 @@
 export type Offset = number;
 
 /**
- * The shared attributes between codecs, encoders and decoders.
- */
-export type CodecData = {
-    /** A description for the codec. */
-    description: string;
-    /** The fixed size of the encoded value in bytes, or `null` if it is variable. */
-    fixedSize: number | null;
-    /** The maximum size an encoded value can be in bytes, or `null` if it is variable. */
-    maxSize: number | null;
+ * An object that can encode a value to a `Uint8Array`.
+ */
+export type Encoder<T> = EncoderSize<T> & {
+    /** Encode the provided value and return the encoded bytes directly. */
+    readonly encode: (value: T) => Uint8Array;
+    /**
+     * Writes the encoded value into the provided byte array at the given offset.
+     * Returns the offset of the next byte after the encoded value.
+     */
+    readonly write: (value: T, bytes: Uint8Array, offset: Offset) => Offset;
 };
 
 /**
- * An object that can encode a value to a `Uint8Array`.
+ * Describes the size of an Encoder, either fixed or variable.
  */
-export type Encoder<T> = CodecData & {
-    /** The function that encodes a value into bytes. */
-    encode: (value: T) => Uint8Array;
-};
+export type EncoderSize<T> =
+    | {
+          /** The fixed size of the encoded value in bytes, if applicable. */
+          readonly fixedSize: number;
+      }
+    | {
+          /** Otherwise, a null fixedSize indicates it's a variable size encoder. */
+          readonly fixedSize: null;
```

Comment:
It strikes me as a bit confusing, if these are being exported, to have the `fixedSize` field on a variable sized encoder.

I can see you're using it for some neat assertion management, but I wonder why you chose not to just check for `undefined` instead?

Reply:
It's a fair point. It makes juggling with sizes a bit more complicated when creating composite codecs (like array, tuples, etc.) but at the benefit of offering a cleaner API for the end-user which IMO is more valuable.

I'll work on that in a separate PR. Thanks!

Reply:
Sweet, sounds good!